cpp/lib/src/Zotero.cc

```diff
@@ -160,6 +160,36 @@ bool Web(const Url &zts_server_url, const TimeLimit &time_limit, Downloader::Par
 } // namespace TranslationServer
 
 
+template <typename T>
+void VisitJsonMetadataNodes(const std::string &node_name, const std::shared_ptr<JSON::JSONNode> &node,
+                            const SiteParams &site_params, const T callback)
+{
+    switch (node->getType()) {
+    case JSON::JSONNode::OBJECT_NODE:
+        for (const auto &key_and_node : *JSON::JSONNode::CastToObjectNodeOrDie(node_name, node))
+            VisitJsonMetadataNodes(key_and_node.first, key_and_node.second, site_params, callback);
+
+        break;
+    case JSON::JSONNode::ARRAY_NODE: {
+        for (const auto &element : *JSON::JSONNode::CastToArrayNodeOrDie(node_name, node)) {
+            if (element->getType() != JSON::JSONNode::OBJECT_NODE)
+                continue;
+
+            const auto object_node(JSON::JSONNode::CastToObjectNodeOrDie("array_element", element));
+            for (auto &key_and_node : *object_node)
+                VisitJsonMetadataNodes(key_and_node.first, key_and_node.second, site_params, callback);
+        }
+
+        break;
+    }
+    case JSON::JSONNode::NULL_NODE:
```

Comment:
Handle this in the callback function as we want the functionality to be as generic as possible.